server/src/main/java/org/apache/iotdb/db/mpp/plan/analyze/PartitionCache.java

```diff
@@ -0,0 +1,710 @@
+    Map<String, String> deviceToStorageGroupMap = new HashMap<>();
+    // miss when devicePath contains *
+    for (String devicePath : devicePaths) {
+      if (devicePath.contains("*")) {
+        return deviceToStorageGroupMap;
+      }
+    }
+    // first try to hit storage group in fast-fail way
+    boolean firstTryResult =
+        getDeviceToStorageGroupMap(devicePaths, deviceToStorageGroupMap, new ArrayList<>(), true);
+    if (!firstTryResult) {
+      try {
+        // when local cache not have, then try to fetch all storage group from config node
+        getAllStorageGroupFromConfigNode();
+        // second try to hit storage group with failed devices;
+        List<String> failedDevices = new ArrayList<>();
+        boolean secondTryResult =
+            getDeviceToStorageGroupMap(devicePaths, deviceToStorageGroupMap, failedDevices, false);
+        if (!secondTryResult && isAutoCreate) {
+          // try to auto create storage group of failed device
+          Set<String> storageGroupNamesNeedCreated = new HashSet<>();
+          for (String devicePath : failedDevices) {
+            PartialPath storageGroupNameNeedCreated =
+                MetaUtils.getStorageGroupPathByLevel(
+                    new PartialPath(devicePath), config.getDefaultStorageGroupLevel());
+            storageGroupNamesNeedCreated.add(storageGroupNameNeedCreated.getFullPath());
+          }
+          createNotExistedStorageGroup(storageGroupNamesNeedCreated);
+          // third try to hit cache
+          boolean thirdTryResult =
+              getDeviceToStorageGroupMap(
+                  devicePaths, deviceToStorageGroupMap, new ArrayList<>(), true);
+          if (!thirdTryResult) {
+            throw new StatementAnalyzeException("Failed to get Storage Group Map in three try.");
+          }
+        }
+      } catch (TException | MetadataException | IOException e) {
+        throw new StatementAnalyzeException(
+            "An error occurred when executing getDeviceToStorageGroup():" + e.getMessage());
+      }
+    }
+    return deviceToStorageGroupMap;
+  }
+
+  /** get all storage group from confignode and update storage group cache */
+  private void getAllStorageGroupFromConfigNode() throws IOException, TException {
+    try (ConfigNodeClient client =
+        configNodeClientManager.borrowClient(ConfigNodeInfo.partitionRegionId)) {
+      TStorageGroupSchemaResp storageGroupSchemaResp =
+          client.getMatchedStorageGroupSchemas(ROOT_PATH);
+      if (storageGroupSchemaResp.getStatus().getCode()
+          == TSStatusCode.SUCCESS_STATUS.getStatusCode()) {
+        Set<String> storageGroupNames = storageGroupSchemaResp.getStorageGroupSchemaMap().keySet();
+        // update all storage group into cache
+        updateStorageCache(storageGroupNames);
+      }
+    }
+  }
+
+  /** create not existed storage group and update storage group cache */
+  private void createNotExistedStorageGroup(Set<String> storageGroupNamesNeedCreated)
```

Comment:
Suggest to rename this method to `createStorageGroupAndUpdateCache()`

Reply:
Get it